src/gui/DatabaseTabWidget.cpp

```diff
@@ -112,126 +97,116 @@ Database* DatabaseTabWidget::execNewDatabaseWizard()
-            if (!i.value().dbWidget->dbHasKey() && !(pw.isNull() && keyFile.isEmpty())) {
-                // If the database is locked and a pw or keyfile is provided, unlock it
-                i.value().dbWidget->switchToOpenDatabase(i.value().fileInfo.absoluteFilePath(), pw, keyFile);
-            } else {
-                setCurrentIndex(databaseIndex(i.key()));
-            }
+    for (int i = 0, c = count(); i < c; ++i) {
+        auto* dbWidget = databaseWidgetFromIndex(i);
+        Q_ASSERT(dbWidget);
+        if (dbWidget && dbWidget->database()->filePath() == canonicalFilePath) {
+            // switch to existing tab if file is already open
+            setCurrentIndex(indexOf(dbWidget));
             return;
         }
     }
 
-    DatabaseManagerStruct dbStruct;
-
-    // test if we can read/write or read the file
-    QFile file(fileName);
-    if (!file.open(QIODevice::ReadWrite)) {
-        if (!file.open(QIODevice::ReadOnly)) {
-            // can't open
-            emit messageGlobal(tr("Unable to open the database.").append("\n").append(file.errorString()),
-                               MessageWidget::Error);
-            return;
-        } else {
-            // can only open read-only
-            dbStruct.readOnly = true;
-        }
-    }
-    file.close();
-
-    Database* db = new Database();
-    dbStruct.dbWidget = new DatabaseWidget(db, this);
-    dbStruct.fileInfo = fileInfo;
+    auto* dbWidget = new DatabaseWidget(QSharedPointer<Database>::create(filePath), this);
+    addDatabaseTab(dbWidget);
+    updateLastDatabases(filePath);
+}
 
-    insertDatabase(db, dbStruct);
+/**
+ * Add a new database tab containing the given DatabaseWidget
+ * @param filePath
+ */
+void DatabaseTabWidget::addDatabaseTab(DatabaseWidget* dbWidget)
+{
+    auto db = dbWidget->database();
+    Q_ASSERT(db);
 
-    if (dbStruct.readOnly) {
-        emit messageTab(tr("File opened in read only mode."), MessageWidget::Warning);
-    }
+    int index = addTab(dbWidget, "");
+    updateTabName(index);
+    setCurrentIndex(index);
+    toggleTabbar();
 
-    updateLastDatabases(dbStruct.fileInfo.absoluteFilePath());
+    connect(dbWidget, SIGNAL(databaseFilePathChanged(QString,QString)), SLOT(updateTabName()));
+    connect(dbWidget, SIGNAL(closeRequest()), SLOT(closeDatabaseTabFromSender()));
+    connect(dbWidget, SIGNAL(databaseModified()), SLOT(updateTabName()));
+    connect(dbWidget, SIGNAL(databaseSaved()), SLOT(updateTabName()));
+    connect(dbWidget, SIGNAL(databaseUnlocked()), SLOT(updateTabName()));
+    connect(dbWidget, SIGNAL(databaseUnlocked()), SLOT(emitDatabaseLockChanged()));
+    connect(dbWidget, SIGNAL(databaseLocked()), SLOT(updateTabName()));
+    connect(dbWidget, SIGNAL(databaseLocked()), SLOT(emitDatabaseLockChanged()));
 
-    if (!pw.isNull() || !keyFile.isEmpty()) {
-        dbStruct.dbWidget->switchToOpenDatabase(dbStruct.fileInfo.absoluteFilePath(), pw, keyFile);
```

Comment:
This should be in DatabaseWidget

Reply:
This?

        if (!db->isInitialized()) {
            dbWidget->switchToOpenDatabase();
        }

Ok.

Reply:
yup